### PR TITLE
improve I2P support

### LINKED ISSUE #7248: [I2P] unknown error
After #7247 fix applied, another error appears.
It happens because of incorrect initialization of `remaining` variable here:
https://github.com/arvidn/libtorrent/blob/690881fec9cc069976cd3d909254a87c22f43ce0/include/libtorrent/i2p_stream.hpp#L285
When such code is used, `remaining` also contains null character, which prevents this comparison from working correctly:
https://github.com/arvidn/libtorrent/blob/690881fec9cc069976cd3d909254a87c22f43ce0/include/libtorrent/i2p_stream.hpp#L309
When such line is used instead:
`string_view remaining(m_buffer.data(), m_buffer.size() - 1);`
unknown error disappears.
Related: #2530

_libtorrent version (or branch)_: 690881fec9cc069976cd3d909254a87c22f43ce0
_platform/architecture_: Windows x64
_compiler and compiler version_: clang 14.0.0

### PR DESCRIPTION
**Pull request by Vort**
Should fix #7248, #7258, #7259, #7260, #7262 and, probably, #2530.

### CODE REVIEW

**Comment by obscuratus-i2p** [comment]
I built and tested this pull request.  It was working fine.  Using a close adaption of the simple magnet client presented by Vort, I was able to successfully complete a torrent download.  I did not experience the delays noted elsewhere, the client promptly picked up my seeder.  Perhaps this was due to the tracker I used.  I didn't use postman's tracker, I used a ZZZOT open tracker.

File: include/libtorrent/i2p_stream.hpp [hunk 1]
```diff
@@ -193,7 +193,7 @@ struct i2p_stream : proxy_base
 
 		// send hello command
 		m_state = read_hello_response;
-		static const char cmd[] = "HELLO VERSION MIN=3.0 MAX=3.0\n";
+		static const char cmd[] = "HELLO VERSION MIN=3.1 MAX=3.1\n";
```
Comment (arvidn) [inline include/libtorrent/i2p_stream.hpp:196]: is there a new feature this update depend on that's only available since SAM version 3.1?

File: include/libtorrent/i2p_stream.hpp [hunk 2]
```diff
@@ -277,7 +277,7 @@ struct i2p_stream : proxy_base
 				break;
 		}
 
-		string_view remaining(m_buffer.data(), m_buffer.size());
+		string_view remaining(m_buffer.data(), m_buffer.size() - 1);
```
Comment (arvidn) [inline include/libtorrent/i2p_stream.hpp:280]: Fixing it this way makes me a but uneasy. Especially by not also ensuring that the character being truncated is in fact the new-line. It would seem a bit more logical to truncate the last character of `m_buffer` on line 241, instead of adding the null terminator. I believe this buffer used to be treated as null terminated.

File: include/libtorrent/i2p_stream.hpp [hunk 3]
```diff
@@ -427,8 +427,8 @@ struct i2p_stream : proxy_base
 		TORRENT_ASSERT(m_magic == 0x1337);
 		m_state = read_session_create_response;
 		char cmd[400];
-		int size = std::snprintf(cmd, sizeof(cmd), "SESSION CREATE STYLE=STREAM ID=%s DESTINATION=TRANSIENT\n"
-			, m_id);
+		static const char options[] = "inbound.quantity=3 outbound.quantity=3 inbound.length=3 outbound.length=3";
```
Comment (arvidn) [inline include/libtorrent/i2p_stream.hpp:430]: is it important to configure the inbound and outbound channels here? I would expect that by leaving these options out, there may be a global configuration that takes effect instead.

File: src/http_connection.cpp [hunk 4]
```diff
@@ -385,6 +385,7 @@ void http_connection::start(std::string const& hostname, int port
 				, std::bind(&http_connection::on_resolve
 				, me, _1, _2));
 		}
+		m_hostname = hostname;
```
Comment (arvidn) [inline src/http_connection.cpp:388]: I don't follow this. Why is this change necessary?

File: src/torrent_info.cpp [hunk 5]
```diff
@@ -1747,6 +1747,10 @@ namespace {
 		e.source = source;
 		m_urls.push_back(e);
 
+#if TORRENT_USE_I2P
+		if (is_i2p_url(e.url)) m_flags |= i2p;
```
Comment (arvidn) [inline src/torrent_info.cpp:1751]: I believe it should be pretty easy to add a unit test for this, which I think would be good. Also, what do you think of checking `m_flags & i2p` as the first condition, to avoid calling `is_i2p_url()` when the flag has already been set.

**Review by arvidn — COMMENTED** [review]
sorry for the delay in reviewing this.

File: include/libtorrent/i2p_stream.hpp (hunk 1 above)
Comment (Vort) [reply include/libtorrent/i2p_stream.hpp:196]: https://geti2p.net/en/docs/api/samv3:
_SAM 3.1 is the recommended minimum SAM implementation because of its support for better signature types than SAM 3.0._
I guess @zzzi2p can explain better why 3.1 is needed (#7258).

File: include/libtorrent/i2p_stream.hpp (hunk 3 above)
Comment (Vort) [reply include/libtorrent/i2p_stream.hpp:430]: Default values for Java I2P and i2pd are different.
So to have similar behaviour it is better to specify values directly.
https://geti2p.net/en/docs/api/samv3:
_Please specify the tunnel quantities in the SESSION CREATE message to get consistent performance with the Java I2P and i2pd routers, using the options e.g. inbound.length=3 outbound.length=3._

File: src/http_connection.cpp (hunk 4 above)
Comment (Vort) [reply src/http_connection.cpp:388]: This is how this block of code was implemented before commit 592544f30f587294351bafce3dee8f53ff7cf03e.
This commit broke I2P support so most safe change in my opinion was to just revert it.

Probably, there is better way of fixing this problem, but I don't perfectly understand how this block of code works to make it better than it was before.

File: include/libtorrent/i2p_stream.hpp (hunk 2 above)
Comment (Vort) [reply include/libtorrent/i2p_stream.hpp:280]: Ok, I will look if I can implement it better.

File: src/http_connection.cpp (hunk 4 above)
Comment (arvidn) [reply src/http_connection.cpp:388]: so moving the assignment to `m_hostname` breaks i2p support?
by manual inspection, they look like they ought to have identical behavior

File: include/libtorrent/i2p_stream.hpp (hunk 3 above)
Comment (arvidn) [reply include/libtorrent/i2p_stream.hpp:430]: I see. is there a good reason to not include these directly in the format string?

Comment (jiigen) [reply include/libtorrent/i2p_stream.hpp:430]: Keeping those out of the format string has been a choice of mine when i first tried that. The reason is that those should be parameters that should be available to set when creating the connection by the client.

File: src/http_connection.cpp (hunk 4 above)
Comment (Vort) [reply src/http_connection.cpp:388]: Commit 592544f30f587294351bafce3dee8f53ff7cf03e puts assignment between `else` and `if`, which executes `if` regardless of what happens inside `#if TORRENT_USE_I2P` block.

File: include/libtorrent/i2p_stream.hpp (hunk 3 above)
Comment (Vort) [reply include/libtorrent/i2p_stream.hpp:430]: Yes, I made it like that because this is how it was suggested by @jigen.
Also I worried about too long strings not fitting in text editor.
It is possible to solve it by breaking string in two parts.
I was thinking about such feature as something non-standard, but I searched more about it, and now I know that adjacent string literals concatenation is perfectly fine method.
I will make this change.

Comment (jiigen) [reply include/libtorrent/i2p_stream.hpp:430]: I even thought that separating made it more readable than a long string.

File: include/libtorrent/i2p_stream.hpp (hunk 1 above)
Comment (zzzi2p) [reply include/libtorrent/i2p_stream.hpp:196]: 3.1 has been supported since 2014 and is the only version supported by i2pd. 3.1 gives you SIGNATURE_TYPE=7 (Ed25519) which gets your users off of insecure DSA_SHA1 signatures. There is no risk of finding a pre-2014 router to talk to, they wouldn't work anyway.

File: include/libtorrent/i2p_stream.hpp (hunk 3 above)
Comment (zzzi2p) [reply include/libtorrent/i2p_stream.hpp:430]: My recommendations about the options, although perhaps better implemented by arvidn later than by Vort in this PR, are here: https://github.com/arvidn/libtorrent/issues/7262#issuecomment-1385362013

Comment (Vort) [reply include/libtorrent/i2p_stream.hpp]: @jiigen look how I made it in 022b53cdc9b42884bd1fb2303be8f81cf5758472.
It is separated and at the same time it is single string.

File: include/libtorrent/i2p_stream.hpp (hunk 2 above)
Comment (Vort) [reply include/libtorrent/i2p_stream.hpp:280]: Looks like null is required at line 247.
I can change `m_dest = &m_buffer[0];` to `m_dest.assign(&m_buffer[0], m_buffer.size());` there.
And then at line 241 `m_buffer[read_pos - 1] = 0;` to `m_buffer.pop_back();`
@arvidn will it be better than ` - 1`?

File: include/libtorrent/i2p_stream.hpp (hunk 3 above)
Comment (jiigen) [reply include/libtorrent/i2p_stream.hpp]: Looks fine to me.

File: src/torrent_info.cpp (hunk 5 above)
Comment (Vort) [reply src/torrent_info.cpp:1751]: I made more tests and found that added check to `torrent_info::add_tracker` fixes #7260 only when fresh torrent is added.
When torrent is started from resume data, `torrent::is_i2p` function returns `false` because `m_torrent_file` have no `m_urls`.
Because of it, I propose to change line
https://github.com/arvidn/libtorrent/blob/32e36984c944e85eacc57c381db4605d6636aa84/src/torrent.cpp#L3148
to `if (is_i2p_url(req.url))` instead.
@arvidn can you check if it is better way to fix #7260 ?

File: include/libtorrent/i2p_stream.hpp (hunk 2 above)
Comment (arvidn) [reply include/libtorrent/i2p_stream.hpp:280]: that's a good point. So this is really just ignoring the null terminator in `m_buffer`.
would you mind adding a `TORRENT_ASSERT(m_buffer[m_buffer.size() - 1] == '\0');`

File: src/http_connection.cpp (hunk 4 above)
Comment (arvidn) [reply src/http_connection.cpp:388]: ouch. I see that now

File: src/torrent_info.cpp (hunk 5 above)
Comment (arvidn) [reply src/torrent_info.cpp:1751]: when I made the comment I mistakenly thought this was done in a loop.
I think setting the flag here makes more sense than checking tracker URLs in the `is_i2p()` function

File: include/libtorrent/i2p_stream.hpp (hunk 2 above)
Comment (Vort) [reply include/libtorrent/i2p_stream.hpp:281]: I added assert in f6d329a4c0df214053edee9d2a11457e89476bfe, but slightly modified it because of compiler complaints.

Comment (arvidn) [reply include/libtorrent/i2p_stream.hpp:281]: at a second thought it would probably have been nicer to make it:
```
TORRENT_ASSERT(m_buffer.back() == '\0');
```
but what you have is fine.

**Review by arvidn — APPROVED** [review]

**Comment by arvidn** [comment]
I intend to land this once CI is green. This should be mentioned in the ChangeLog, but I could add that later

**Comment by arvidn** [comment]
@Vort @jiigen Thanks for your contributions!
